onnxruntime/core/providers/webnn/builders/impl/mha_op_builder.cc

```diff
@@ -0,0 +1,369 @@
+      }
+      value_input = model_builder.GetOperand(input_defs[2]->Name());
+      if (input_query_type == ONNX_NAMESPACE::TensorProto_DataType_FLOAT16) {
+        common_options.set("label", node.Name() + "_/MHA/preprocess/cast/value_input");
+        value_input = model_builder.GetBuilder().call<emscripten::val>("cast", value_input, emscripten::val("float32"),
+                                                                       common_options);
+      }
+
+      ORT_RETURN_IF_NOT(GetShape(*input_defs[2], input_v_shape, logger), "Cannot get value shape");
+      const auto v_rank = input_v_shape.size();
+      if (v_rank == 3) {  // Value with shape (batch_size, kv_sequence_length, v_hidden_size)
+        v_reshape_skip = false;
+      } else {  // past_value with shape (batch_size, num_heads, kv_sequence_length, head_size)
+        v_reshape_skip = true;
+      }
+    }
+  } else {  // packed QKV with shape (batch_size, kv_sequence_length, num_heads, 3, head_size)
+    kv_sequence_length = SafeInt<uint32_t>(input_q_shape[2]);
+    head_size = SafeInt<uint32_t>(input_q_shape[4]);
+    hidden_size = SafeInt<uint32_t>(num_heads * head_size);
+    k_reshape_skip = false;
+    v_reshape_skip = false;
+    options.set("axis", 3);
+    emscripten::val output_array = model_builder.GetBuilder().call<emscripten::val>("split", query_input, 3, options);
+    query_input = output_array[0];
+    key_input = output_array[1];
+    value_input = output_array[2];
+  }
+
+  emscripten::val attention_bias = emscripten::val::undefined();
+  if (!TensorExists(input_defs, 5)) {
+    attention_bias = model_builder.GetOperand(input_defs[5]->Name());
+    if (input_query_type == ONNX_NAMESPACE::TensorProto_DataType_FLOAT16) {
+      common_options.set("label", node.Name() + "_/MHA/preprocess/cast/attention_bias");
+      attention_bias = model_builder.GetBuilder().call<emscripten::val>("cast", attention_bias,
+                                                                        emscripten::val("float32"), common_options);
+    }
+  }
+
+  batch_size = SafeInt<uint32_t>(input_q_shape[0]);
+  sequence_length = SafeInt<uint32_t>(input_q_shape[1]);
+
+  const float scale_value = helper.Get("scale", 1 / sqrt(static_cast<float>(head_size)));
+
+  const std::vector<uint32_t> reshape_output_shape = {batch_size, sequence_length, hidden_size};
+  const std::vector<uint32_t> q_reshape_tensor_shape = {batch_size, sequence_length, num_heads, head_size};
+  const std::vector<uint32_t> reshape_tensor_shape = {batch_size, kv_sequence_length, num_heads, head_size};
+
+  // query_input -> reshape(B,S,N,H) -> transpose(B,N,S,H) -> new_query
+  common_options.set("label", node.Name() + "_/MHA/query/reshape");
+  emscripten::val reshaped_query = model_builder.GetBuilder().call<emscripten::val>(
+      "reshape", query_input, emscripten::val::array(q_reshape_tensor_shape), common_options);
+
+  options.set("permutation", emscripten::val::array(std::vector<uint32_t>({0, 2, 1, 3})));
```

Comment:
Now this `options` has contained `axis` member, you'd better remove it or use a new one.